adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -274,47 +314,83 @@ def _final_callback(self):
+
+    subclasses = []
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls.subclasses.append(cls)
+
+    def step(self, *args, **kwargs):
         if not self._adp.require_backward_grad_sync:
             return
         scale = self._accum_scale * self._accum_count
         initial_lr = [pg["lr"] for pg in self._optimizer.param_groups]
-        for sqr, var, pg in zip(self._state["sqr_avg"], self._state["var_avg"],
-                                self._optimizer.param_groups):
-            sqr, var = max(float(sqr), 0.0), max(float(var), 1e-6)
-            pg["lr"] = (var + sqr) / (var / scale + sqr) * pg["lr"]
+        lr_factors = self._calculate_lr_factors(scale)
+        for lr_factor, pg in zip(lr_factors, self._optimizer.param_groups):
+            pg["lr"] = lr_factor * pg["lr"]
         self._orig_optimizer_step(*args, **kwargs)
         for lr, pg in zip(initial_lr, self._optimizer.param_groups):
             pg["lr"] = lr
         self._state["progress"] += self.gain(scale)
         self._reset_accumulation()
 
-    def zero_grad(self, *args, **kwargs):
-        if self._should_zero_grad:
-            self._orig_optimizer_zero_grad(*args, **kwargs)
-        else:
-            warnings.warn("skipping zero_grad for accumulated gradient")
 
-    def patch_optimizer(self):
-        """
-        Monkey-patch the optimizer's step function with :meth:`AdaScale.step`.
-        """
-        # TODO: detect if the optimizer has already been patched.
+class AdaScale(ScalingRuleBase):
+    """
+    Implements the AdaScale_ algorithm for scaling the learning rate for
+    distributed and large batch size training. Can be used in combination with
+    ``torch.nn.parallel.DistributedDataParallel`` and ``torch.optim.SGD``.
 
-        @functools.wraps(self._optimizer.step)
-        def step_wrapper(optim, *args, **kwargs):
-            return self.step(*args, **kwargs)
+    .. code-block:: python
 
-        @functools.wraps(self._optimizer.zero_grad)
-        def zero_wrapper(optim, *args, **kwargs):
-            return self.zero_grad(*args, **kwargs)
-        self._optimizer.step = MethodType(step_wrapper, self._optimizer)
-        self._optimizer.zero_grad = MethodType(zero_wrapper, self._optimizer)
+        optim = torch.optim.SGD(model, lr=0.001)
+        model = DistributedDataParallel(model, optim, scaling_rule="AdaScale")
+        adascale = model.scaling_rule
+
+        for epoch in ...:
+            for batch in ...:
+                optim.zero_grad()
+                loss = ...
+                loss.backward()
+                adascale.step()
+
+    .. _AdaScale: https://proceedings.icml.cc/static/paper_files/icml/2020/4682-Supplemental.pdf
+    """  # noqa: E501
+
+    def _calculate_lr_factors(self, scale):
+        """Calculate factors to be applied to lr for each parameter group."""
+        lr_factors = []
+        for sqr, var, pg in zip(self._state["sqr_avg"], self._state["var_avg"],
```

Comment:
You don't need `pg` here as its not used in the loop. Because `sqr` and `var` are numpy arrays, you can remove this loop and replace it with numpy math.